server/src/main/java/org/apache/iotdb/db/tools/TsFileRewriteTool.java

```diff
@@ -0,0 +1,478 @@
+          TSDataType.INT64);
+  protected Decoder valueDecoder;
+
+  /** PartitionId -> TsFileIOWriter */
+  protected Map<Long, TsFileIOWriter> partitionWriterMap;
+
+  /** Maximum index of plans executed within this TsFile. */
+  protected long maxPlanIndex = Long.MIN_VALUE;
+
+  /** Minimum index of plans executed within this TsFile. */
+  protected long minPlanIndex = Long.MAX_VALUE;
+
+  /**
+   * Create a file reader of the given file. The reader will read the real data and rewrite to some
+   * new tsFiles.
+   *
+   * @throws IOException If some I/O error occurs
+   */
+  public TsFileRewriteTool(TsFileResource resourceToBeRewritten) throws IOException {
+    oldTsFile = resourceToBeRewritten.getTsFile();
+    String file = oldTsFile.getAbsolutePath();
+    reader = new TsFileSequenceReader(file);
+    partitionWriterMap = new HashMap<>();
+    if (FSFactoryProducer.getFSFactory().getFile(file + ModificationFile.FILE_SUFFIX).exists()) {
+      oldModification = (List<Modification>) resourceToBeRewritten.getModFile().getModifications();
+      modsIterator = oldModification.iterator();
+      fileModificationMap = new HashMap<>();
+    }
+  }
+
+  /**
+   * Rewrite an old file to the latest version
+   *
+   * @param resourceToBeRewritten the tsfile which to be rewrite
+   * @param rewrittenResources the rewritten files
+   */
+  public static void rewriteTsFile(
+      TsFileResource resourceToBeRewritten, List<TsFileResource> rewrittenResources)
+      throws IOException, WriteProcessException {
+    try (TsFileRewriteTool rewriteTool = new TsFileRewriteTool(resourceToBeRewritten)) {
+      rewriteTool.parseAndRewriteFile(rewrittenResources);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.reader.close();
+  }
+
+  /**
+   * Parse the old files and generate some new files according to the time partition interval.
+   *
+   * @throws IOException WriteProcessException
+   */
+  @SuppressWarnings({"squid:S3776", "deprecation"}) // Suppress high Cognitive Complexity warning
+  public void parseAndRewriteFile(List<TsFileResource> rewrittenResources)
+      throws IOException, WriteProcessException {
+    // check if the TsFile has correct header
+    if (!fileCheck()) {
+      return;
+    }
+    int headerLength = TSFileConfig.MAGIC_STRING.getBytes().length + Byte.BYTES;
+    reader.position(headerLength);
+    // start to scan chunks and chunkGroups
+    List<List<PageHeader>> pageHeadersInChunkGroup = new ArrayList<>();
+    List<List<ByteBuffer>> pageDataInChunkGroup = new ArrayList<>();
+    List<List<Boolean>> needToDecodeInfoInChunkGroup = new ArrayList<>();
+    byte marker;
+    List<MeasurementSchema> measurementSchemaList = new ArrayList<>();
+    String lastChunkGroupDeviceId = null;
+    try {
+      while ((marker = reader.readMarker()) != MetaMarker.SEPARATOR) {
+        switch (marker) {
+          case MetaMarker.CHUNK_HEADER:
+          case MetaMarker.ONLY_ONE_PAGE_CHUNK_HEADER:
```

Comment:
Not sure, do we need deal with `CHUNK_HEADER`?

Reply:
In the new tsfile format, CHUNK_HEADER and ONLY_ONE_PAGE_CHUNK_HEADER both indicts the start of one chunk. @HTHou is an expert on tsfile format, please take a look at.

Reply:
if "only one page in the chunk", then there is no pageStatistics in the page header. 

Line 170 and 171 has indicated that in PageHeader's deserialize method: 
```
PageHeader pageHeader =
                  reader.readPageHeader(dataType, header.getChunkType() == MetaMarker.CHUNK_HEADER);
```